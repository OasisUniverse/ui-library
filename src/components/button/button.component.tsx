import React, { FC, useMemo } from 'react';
import styles from './button.module.scss';

export enum ButtonSizes {
    Large = 'large',
    Medium = 'medium',
    Small = 'small',
}

export enum ButtonType {
    Default = 'default',
    Secondary = 'secondary',
    Outlined = 'outlined',
}

export interface ButtonProps {
    className?: string;
    size?: ButtonSizes;
    type?: ButtonType;
    text?: string;
    onClick: (event?: React.MouseEvent<HTMLButtonElement>) => void;
    icon?: JSX.Element;
    reverseItems?: boolean;
    disabled?: boolean;
}

export const Button: FC<ButtonProps> = ({
    className = '',
    size = ButtonSizes.Medium,
    type = ButtonType.Default,
    text,
    onClick,
    icon,
    reverseItems = false,
    disabled = false,
}) => {
    const buttonClassName = useMemo(
        () =>
            `${styles.button} ${type ? styles[type] : ''} ${size ? styles[size] : ''} ${
                icon && text ? styles.buttonWithIcon : ''
            } ${reverseItems ? styles.reverseItems : ''} ${className} ${disabled ? styles.disabled : ''}`,
        [type, size, className, disabled, text, icon, reverseItems],
    );
    return (
        <button className={buttonClassName} onClick={onClick} disabled={disabled}>
            {icon && icon}
            {text && text}
        </button>
    );
};
