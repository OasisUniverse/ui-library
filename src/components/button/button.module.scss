@import '../../styles/colors.scss';
@import '../../styles/fonts';

.button {
	width: 100%;
	display: flex;
	flex-direction: row;
	justify-content: center;
	outline: none;
	border: none;
	align-items: center;
	box-sizing: border-box;
	border-radius: 12px;
	font-weight: 500;
	cursor: pointer;
	transition: 0.2s ease all;
	font-family: Inter, sans-serif;
	user-select: none;
	
	&.buttonWithIcon {
		justify-content: space-around;
	}
	
	&.reverseItems {
		flex-direction: row-reverse;
	}
	
	&.default {
		color: white;
		background: color(primary, dark);
		
		&:hover {
			box-shadow: 0 0 15px color(primary, dark);
		}
		
		&:active {
			box-shadow: 0 0 15px color(primary, dark);
			background: color(background, content);
		}
		
		&:disabled {
			color: white;
			background: color(statuses, muted);
		}
	}
	
	&.secondary {
		color: white;
		background: color(background, content);
		border: 2px solid color(primary, dark);
		
		&:hover {
			box-shadow: 0 0 15px color(primary, dark);
		}
		
		&:active {
			box-shadow: inset 0 0 15px color(background, main), 0 0 10px color(primary, dark);
		}
		
		&:disabled {
			color: white;
			background: color(inneres, placeholder);
			border: 2px solid color(statuses, muted);
		}
	}
	
	&.outlined {
		color: color(accent, dark);
		background: transparent;
		border: 1px solid color(accent, dark);
		
		&:hover {
			box-shadow: 0 0 15px color(accent, dark);
		}
		
		&:active {
			background: color(accent, dark);
			color: white;
		}
		
		&:disabled {
			color: color(statuses, muted);
			background: transparent;
			border: 1px solid color(statuses, muted);
		}
	}
	
	&.disabled:hover {
		box-shadow: none;
		cursor: not-allowed;
	}
	
	&.small {
		justify-content: center;
		align-items: center;
		height: 64px;
		width: 64px;
	}
	
	&.medium {
		font-size: 16px;
		line-height: 16px;
		padding: 12px 24px;
		height: 40px;
	}
	
	&.large {
		font-size: 24px;
		height: 64px;
	}
}