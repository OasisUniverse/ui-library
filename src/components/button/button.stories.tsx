import React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import { Button, ButtonProps, ButtonSizes, ButtonType } from './button.component';
import { buttonTestMock } from '../../__mocks__/button.test.mock';

export default {
    title: 'Button',
    component: Button,
} as Meta;

const Template: Story<ButtonProps> = (args: ButtonProps) => <Button {...args} />;

export const Initial = Template.bind({});

Initial.args = {
    text: 'Text',
    size: ButtonSizes.Medium,
    type: ButtonType.Default,
    onClick: () => null,
};

const presentationStyle = {
    width: 'auto',
    background: '#1a1a22',
    padding: '50px',
    display: 'grid',
    gridTemplateColumns: 'repeat(3, 1fr)',
    gridTemplateRows: 'auto',
    gap: '50px',
};

export const Presentation = (): JSX.Element => (
    <div style={presentationStyle}>
        {Array.from(buttonTestMock, ([, props]) => (
            <Button {...props} />
        ))}
    </div>
);
