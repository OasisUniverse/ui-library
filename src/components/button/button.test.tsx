import React from 'react';
import { Button, ButtonProps } from './button.component';
import Enzyme, { mount, shallow, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { buttonTestMock } from '../../__mocks__/button.test.mock';
import LoadingSpinner from '../../shared/svg/loading-spinner.icon';
import InputCrossIcon from '../../shared/svg/input-cross.icon';

const defaultProps: ButtonProps = {
    onClick: jest.fn(),
};

Enzyme.configure({ adapter: new Adapter() });

let button: ShallowWrapper;

describe('<Button /> test list', () => {
    beforeAll(() => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('default')} />);
        jest.clearAllMocks();
    });
    it('not disabled', () => {
        expect((button.props() as ButtonProps).disabled).toBeFalsy();
        console.log(button.props());
    });
    it('disabled', () => {
        const button = shallow(<Button onClick={jest.fn()} {...buttonTestMock.get('default-disabled')} />);
        expect((button.props() as ButtonProps).disabled).toBeTruthy();
        expect(button.hasClass('disabled')).toBeTruthy();
    });
    it('has a large size', () => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('default-large')} />);
        expect(button.hasClass('large')).toBeTruthy();
    });
    it('has an icon inside', () => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('default-small')} />);
        expect(button.childAt(0).exists()).toBeTruthy();
    });
    it('has an icon and text inside', () => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('icon-and-text')} />);
        expect(button.childAt(0).exists()).toBeTruthy();
        expect(button.childAt(1).exists()).toBeTruthy();
        expect(button.hasClass('buttonWithIcon')).toBeTruthy();
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('icon-and-text-reversed')} />);
        expect(button.hasClass('reverseItems')).toBeTruthy();
    });
    it('has every size', () => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('default')} />);
        expect(button.hasClass('medium')).toBeTruthy();
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('default-large')} />);
        expect(button.hasClass('large')).toBeTruthy();
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('default-small')} />);
        expect(button.hasClass('small')).toBeTruthy();
    });
    it('has every type', () => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('default')} />);
        expect(button.hasClass('default')).toBeTruthy();
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('secondary')} />);
        expect(button.hasClass('secondary')).toBeTruthy();
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('outlined')} />);
        expect(button.hasClass('outlined')).toBeTruthy();
    });
    it('is loading', () => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('loading-medium')} />);
        expect(button.hasClass('loading')).toBeTruthy();
        expect(button.find(LoadingSpinner).exists()).toBeTruthy();
    });
    it('is link and not loading', () => {
        button = shallow(<Button {...defaultProps} {...buttonTestMock.get('not-loading-link')} />);
        expect(button.find('a').exists()).toBeTruthy();
        expect(button.find(InputCrossIcon).exists()).toBeTruthy();
        expect(button.find('a').text()).toEqual('<InputCrossIcon />some text1');
        expect(button.hasClass('buttonWithIcon')).toBeTruthy();
    });
    it('is disabled link', () => {
        let wrapper = mount(<Button onClick={jest.fn()} {...buttonTestMock.get('disabled-link')} disabled={true} />);
        const link = wrapper.find('a');
        expect((wrapper.props() as ButtonProps).disabled).toBeTruthy();
        expect(link.hasClass('disabled')).toBeTruthy();
        wrapper = mount(<Button onClick={jest.fn()} {...buttonTestMock.get('disabled-link')} disabled={false} />);
        expect((wrapper.props() as ButtonProps).disabled).toBeFalsy();
        expect(link.hasClass('medium')).toBeTruthy();
        expect(link.hasClass('default')).toBeTruthy();
        expect(link.hasClass('button')).toBeTruthy();
    });
});
