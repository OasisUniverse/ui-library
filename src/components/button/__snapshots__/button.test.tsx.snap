// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> test list render all list 1`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 2`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 3`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 4`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 5`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 6`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 7`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 8`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 9`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 10`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 11`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 12`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 13`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 14`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 15`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 16`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 17`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 18`] = `ShallowWrapper {}`;

exports[`<Button /> test list render all list 19`] = `ShallowWrapper {}`;
